kernel/src/grant.rs

```diff
@@ -153,6 +169,166 @@ impl<'a, T: 'a + ?Sized> DerefMut for GrantMemory<'a, T> {
     }
 }
 
+/// This GrantUpcallTable object provides a handle to access Upcalls stored on
+/// behalf of a particular grant/driver.
+///
+/// Capsules gain access to a GrantUpcallTable object by calling
+/// `Grant::enter()`. From there, they can schedule upcalls. No other access to
+/// upcalls is provided.
+///
+/// It is expected that this type will only exist as a short-lived stack
+/// allocation, so its size is not a significant concern.
+pub struct GrantUpcallTable<'a> {
+    /// The mutable reference to the actual object type stored in the grant.
+    upcalls: &'a [SavedUpcall],
+
+    /// We need to keep track of the driver number so we can properly identify
+    /// the Upcall that is called. We need to keep track of its source so we can
+    /// remove it if the Upcall is unsubscribed.
+    driver_num: usize,
+
+    /// A reference to the process that these upcalls are for. This is used for
+    /// actually scheduling the upcalls.
+    process: &'a dyn Process,
+}
+
+impl<'a> GrantUpcallTable<'a> {
+    /// Create a `GrantUpcallTable` object to provide a handle for capsules to
+    /// call Upcalls.
+    fn new(
+        upcalls: &'a [SavedUpcall],
+        driver_num: usize,
+        process: &'a dyn Process,
+    ) -> GrantUpcallTable<'a> {
+        Self {
+            upcalls,
+            driver_num,
+            process,
+        }
+    }
+
+    /// Schedule the specified upcall for the process with r0, r1, r2 as
+    /// provided values.
+    ///
+    /// Capsules call this function to schedule upcalls, and upcalls are
+    /// identified by the `subscribe_num`, which must match the subscribe number
+    /// used when the upcall was originally subscribed by a process.
+    /// `subscribe_num`s are indexed starting at zero.
+    pub fn schedule_upcall(&self, subscribe_num: usize, r0: usize, r1: usize, r2: usize) -> bool {
```

Comment:
Should this return a `Result<(), Error>` rather than a bool?

It's a bit annoying from a verbosity perspective, as it results in a boilerplate `upcalls.schedule_upcall(..)` -> `upcalls.schedule_upcall(..).ok()` for most callers to silence the ignored `Result`, but it does seem clearer than the `bool`, and there is some benefit to asking capsule authors to consider briefly whether they care about the process not actually being notified.

Reply:
I just ported the existing interface for `schedule()`. I think you're right that a `Result` would be better, but I might prefer for that to be submitted as a follow up PR, since it would require updating all the capsules again, and the current content of this PR has been pretty heavily reviewed.

Reply:
Does a capsule have any recourse if a schedule() fails? I'm guessing that is why this has escaped the resultifying this long, I'm not sure what a capsule could possibly do that would be worth it. Has any capsule ever even handled the bool?

I'm not saying we want to stick with the c convention, but we also probably don't want `upcalls.schedule().map_err(debug!("i have no idea what happened"))`.

Reply:
I don't have a compelling in-tree example I don't think, but my guess was that it was more likely that some capsule might need to do something extra if it thinks the process has died (e.g. check in some other way if the process is truly dead after the subscribe fails and choose to power off the associated peripheral).

It doesn't need the whole `map_err` bit, literally appending `.ok()` as shown is sufficient; that is a "Rust trick" I read somewhere along the way that just turns a `Result` into an `Option` (you lose the `Error` when you do this; Success -> `Some`, any error -> `None`), and the `Option` can be dropped without the compiler yelling.

@hudson-ayers follow-up PR seems fine; should be a pretty mechanical change.

Reply:
Thinking about this a little more, it's impossible for .schedule() to detect a dead process, because to enter a grant the process must be valid and while the grant is entered the process must stay valid. So schedule could fail because:

1. the `subscribe_num` was invalid (i.e. too big)
2. the upcall is null, i.e. the process never called subscribe
3. there is no room in the tasks array for the process

We don't need a persistent runtime check for 1, number 2 is normal operation that capsules aren't expected to defend against (I'm pretty sure), and for 3 its not clear what the capsule should do. Perhaps encapsulating these three cases in an Error type would spur someone to think about this any maybe come up with something.

And `ok()` is the clear tool for capsule authors to use, but I was just suggesting that if schedule() returns a proper result and a capsule author is thinking about what to do with the error case (i.e. "why does the kernel return me an error if I am not expected to do anything with it?"), they likely won't come up with anything and might write a debug statement instead.

Reply:
Opened #2656

Reply:
If `schedule` cannot detect a dead process, what are the error cases?

It seems like this should return nothing. A `Result` is just going to add a lot of boilerplate that doesn't do anything.